delivery/common/jwt/lib/jwt.js

```diff
@@ -0,0 +1,128 @@
+            if (jwtPayload.nbf && jwtPayload.nbf > clockTimestamp + (this.jwtOptions.clockTolerance || 0)) throw new Error("JWT not active");
+        }
+        if (!jwtHeader.alg) throw new Error("JWT malformed: expected alg field in JWT header");
+        if ("NONE" === jwtHeader.alg.toUpperCase()) return {
+            header: jwtHeader,
+            payload: jwtPayload
+        };
+        if (!(null === (_a = this.algorithms) || void 0 === _a ? void 0 : _a.includes(jwtHeader.alg.toUpperCase()))) throw new Error(`${jwtHeader.alg} is not supported at the moment`);
+        for (const cryptoKey of keys) {
+            if (await this.validateSignature(jwtParts, jwtHeader.alg.toUpperCase(), cryptoKey)) return {
+                header: jwtHeader,
+                payload: jwtPayload
+            };
+        }
+        throw new Error("JWT token signature verification failed!");
+    }
+    validateOptionTypes() {
+        if (void 0 !== this.jwtOptions.issuer && ("string" != typeof this.jwtOptions.issuer || 0 === this.jwtOptions.issuer.trim().length)) throw new Error("Invalid jwtOptions: issuer must be non empty string");
+        if (void 0 !== this.jwtOptions.audience && ("string" != typeof this.jwtOptions.audience || 0 === this.jwtOptions.audience.trim().length)) throw new Error("Invalid jwtOptions: audience must be non empty string");
+        if (void 0 !== this.jwtOptions.subject && ("string" != typeof this.jwtOptions.subject || 0 === this.jwtOptions.subject.trim().length)) throw new Error("Invalid jwtOptions: subject must be non empty string");
+        if (void 0 === this.jwtOptions.ignoreExpiration) this.jwtOptions.ignoreExpiration = !0; else if ("boolean" != typeof this.jwtOptions.ignoreExpiration) throw new Error("Invalid jwtOptions: ignoreExpiration must be boolean");
+        if (void 0 === this.jwtOptions.ignoreNotBefore) this.jwtOptions.ignoreNotBefore = !0; else if ("boolean" != typeof this.jwtOptions.ignoreNotBefore) throw new Error("Invalid jwtOptions: ignoreNotBefore must be boolean");
+        if (void 0 === this.jwtOptions.clockTolerance) this.jwtOptions.clockTolerance = 60; else if ("number" != typeof this.jwtOptions.clockTolerance) throw new Error("Invalid jwtOptions: clockTimestamp must be number");
+    }
+    async validateSignature(jwtParts, alg, cryptoKey) {
+        switch (alg) {
+          case "RS512":
+          case "RS384":
+          case "RS256":
+            return await crypto.subtle.verify({
+                name: "RSASSA-PKCS1-v1_5"
+            }, cryptoKey, base64url.decode(jwtParts[2], "Uint8Array").buffer, (new TextEncoder).encode(`${jwtParts[0]}.${jwtParts[1]}`));
+
+          case "HS512":
+          case "HS384":
+          case "HS256":
+            return await crypto.subtle.verify({
+                name: "HMAC"
+            }, cryptoKey, base64url.decode(jwtParts[2], "Uint8Array").buffer, (new TextEncoder).encode(`${jwtParts[0]}.${jwtParts[1]}`));
```

Comment:
Would it be possible to reuse the original string here, instead of concatenating the header and payload? 

Something like
```javascript
base64JWTToken.substring(0, jwtParts[0].length + 1 + jwtParts[1].length)
```

Either computing that in `validate()` and passing it in or passing the array in to this function.

Reply:
@evan-hughes : This would make an extra operation in the validateSignature function. 
The original base64url JWTToken is already split into parts in validate function and the parts (as array) is passed for later use in validateSignature function. 
So validateSignature already contains jwtToken in parts (i.e header, payload and signature ) that can be used for signature verification.
Any thoughts what would be the benefit of again splitting the original base64JWTToken to obtain the same in validateSignature function?

Reply:
I'm suggesting that the following string creation:
```javascript 
(new TextEncoder).encode(`${jwtParts[0]}.${jwtParts[1]}`)
```

be replaced with the `base64JWTToken.substring(...)` listed above. 

AFAIU v8 creates substrings without copying the underlying string, so the `base64JWTToken.substring(...)` allows us to avoid an allocation and two writes, which is necessary with the ``${jwtParts[0]}.${jwtParts[1]}``. 

I don't think those allocations/copies are hugely expensive, so leaving them is not a significant problem.

Reply:
Updated with the mentioned changes